# Problema

Predecir el coste del seguro

# El set de datos

* age: age of primary beneficiary

* sex: insurance contractor gender, female, male

* bmi: Body mass index, providing an understanding of body, weights that are relatively high or low relative to height,
objective index of body weight (kg / m ^ 2) using the ratio of height to weight, ideally 18.5 to 24.9

* children: Number of children covered by health insurance / Number of dependents

* smoker: Smoking

* region: the beneficiary's residential area in the US, northeast, southeast, southwest, northwest.

* charges: Individual medical costs billed by health insurance



In [31]:
# imports
import pandas as pd
from google.colab import drive
drive.mount('/gdrive')

Drive already mounted at /gdrive; to attempt to forcibly remount, call drive.mount("/gdrive", force_remount=True).


In [32]:
ruta = "/content/insurance.csv"
df = pd.read_csv(ruta)

In [33]:
print(df.shape)
df.head()

(1338, 7)


age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

# Objetivo

Generar un model de regresión capaz de predecir el valor del seguro en base a las características del cliente.

* Aplicar las técnicas oportunas de procesamiento de datos

* Valorar diferentes modelos de regresión

* Comparación entre modelos

* Ensemble

* Métricas

* Conclusiones finales

## Implementación

In [34]:
#vemos los valores únicos de cada una de las columnas

df['charges'].unique()



array([16884.924 ,  1725.5523,  4449.462 , ...,  1629.8335,  2007.945 ,
       29141.3603])

In [35]:
import pandas as pd

# verificamos que no hay filas duplicadas, en caso de que haya alguna, la eliminamos

# Verificar si hay filas duplicadas
duplicated_rows = df.duplicated()

# Filtrar el DataFrame original para mostrar solo las filas duplicadas
duplicated_data = df[duplicated_rows]

# Verificar si la fila 581 está duplicada
if 581 in duplicated_data.index:
    # Eliminar la fila específica (fila 581)
    df.drop(581, inplace=True)
    print("La fila 581 fue una fila duplicada y ha sido eliminada.")

# Mostrar las filas duplicadas
print("Filas duplicadas:")
print(duplicated_data)

# Eliminar las filas duplicadas del DataFrame original
df.drop_duplicates(inplace=True)







La fila 581 fue una fila duplicada y ha sido eliminada.
Filas duplicadas:
     age   sex    bmi  children smoker     region    charges
581   19  male  30.59         0     no  northwest  1639.5631


In [36]:
#normalizamos la variable numérica bmi para tenerla a escala y que ningun dato se apodere de otro

import pandas as pd
from sklearn.preprocessing import MinMaxScaler


# Seleccionar la columna 'bmi'
bmi_column = df['bmi']

# Reshape de la columna 'bmi' si es necesario para que sea una matriz 2D
bmi_column = bmi_column.values.reshape(-1, 1)

# Crear un objeto MinMaxScaler
scaler = MinMaxScaler()

# Normalizar los datos
bmi_normalized = scaler.fit_transform(bmi_column)

# Imprimir los resultados de la normalización
print("Valores originales de 'bmi':")
print(bmi_column)
print("\nValores normalizados de 'bmi':")
print(bmi_normalized)

# Reemplazar la columna 'bmi' en el DataFrame original con los datos normalizados
df['bmi'] = bmi_normalized




Valores originales de 'bmi':
[[27.9 ]
 [33.77]
 [33.  ]
 ...
 [36.85]
 [25.8 ]
 [29.07]]

Valores normalizados de 'bmi':
[[0.3212268 ]
 [0.47914985]
 [0.45843422]
 ...
 [0.56201238]
 [0.26472962]
 [0.35270379]]


In [37]:
df.head()

age     sex       bmi  children smoker     region      charges
0   19  female  0.321227         0    yes  southwest  16884.92400
1   18    male  0.479150         1     no  southeast   1725.55230
2   28    male  0.458434         3     no  southeast   4449.46200
3   33    male  0.181464         0     no  northwest  21984.47061
4   32    male  0.347592         0     no  northwest   3866.85520

In [38]:
##normalizar age


import pandas as pd
from sklearn.preprocessing import MinMaxScaler


# Seleccionar la columna 'age'
age_column = df['age']

# Reshape de la columna 'bmi' si es necesario para que sea una matriz 2D
age_column = age_column.values.reshape(-1, 1)

# Crear un objeto MinMaxScaler
scaler = MinMaxScaler()

# Normalizar los datos
age_normalized = scaler.fit_transform(age_column)

# Imprimir los resultados de la normalización
print("Valores originales de 'age':")
print(age_column)
print("\nValores normalizados de 'age':")
print(age_normalized)

# Reemplazar la columna 'bmi' en el DataFrame original con los datos normalizados
df['age'] = age_normalized


Valores originales de 'age':
[[19]
 [18]
 [28]
 ...
 [18]
 [21]
 [61]]

Valores normalizados de 'age':
[[0.02173913]
 [0.        ]
 [0.2173913 ]
 ...
 [0.        ]
 [0.06521739]
 [0.93478261]]


In [39]:
df.head()

age     sex       bmi  children smoker     region      charges
0  0.021739  female  0.321227         0    yes  southwest  16884.92400
1  0.000000    male  0.479150         1     no  southeast   1725.55230
2  0.217391    male  0.458434         3     no  southeast   4449.46200
3  0.326087    male  0.181464         0     no  northwest  21984.47061
4  0.304348    male  0.347592         0     no  northwest   3866.85520

In [40]:
#en las variables categoricas como sex, le ponemos una label a cada una de las opciones
from sklearn.preprocessing import LabelEncoder

# Crear un objeto LabelEncoder
le = LabelEncoder()

# Ajustar y transformar la variable 'sex'
df['sex'] = le.fit_transform(df['sex'])

# Mostrar el mapeo entre categorías y números
print("Mapeo de categorías a números:")
for i, category in enumerate(le.classes_):
    print(f"{category}: {i}")

# Mostrar las primeras filas del DataFrame con la variable 'sex' codificada
print("\nDataFrame con 'sex' codificada:")
print(df.head())


Mapeo de categorías a números:
female: 0
male: 1

DataFrame con 'sex' codificada:
        age  sex       bmi  children smoker     region      charges
0  0.021739    0  0.321227         0    yes  southwest  16884.92400
1  0.000000    1  0.479150         1     no  southeast   1725.55230
2  0.217391    1  0.458434         3     no  southeast   4449.46200
3  0.326087    1  0.181464         0     no  northwest  21984.47061
4  0.304348    1  0.347592         0     no  northwest   3866.85520


In [43]:
from sklearn.preprocessing import LabelEncoder

# Crear un objeto LabelEncoder
le = LabelEncoder()

# Ajustar y transformar la variable 'sex'
df['smoker'] = le.fit_transform(df['smoker'])

# Mostrar el mapeo entre categorías y números
print("Mapeo de categorías a números:")
for i, category in enumerate(le.classes_):
    print(f"{category}: {i}")

# Mostrar las primeras filas del DataFrame con la variable 'smoker' codificada
print("\nDataFrame con 'sex' codificada:")
print(df.head())

Mapeo de categorías a números:
no: 0
yes: 1

DataFrame con 'sex' codificada:
        age  sex       bmi  children  smoker     region      charges
0  0.021739    0  0.321227         0       1  southwest  16884.92400
1  0.000000    1  0.479150         1       0  southeast   1725.55230
2  0.217391    1  0.458434         3       0  southeast   4449.46200
3  0.326087    1  0.181464         0       0  northwest  21984.47061
4  0.304348    1  0.347592         0       0  northwest   3866.85520


In [44]:
from sklearn.preprocessing import LabelEncoder

# Crear un objeto LabelEncoder
le = LabelEncoder()

# Ajustar y transformar la variable 'sex'
df['region'] = le.fit_transform(df['region'])

# Mostrar el mapeo entre categorías y números
print("Mapeo de categorías a números:")
for i, category in enumerate(le.classes_):
    print(f"{category}: {i}")

# Mostrar las primeras filas del DataFrame con la variable 'region' codificada
print("\nDataFrame con 'sex' codificada:")
print(df.head())

Mapeo de categorías a números:
northeast: 0
northwest: 1
southeast: 2
southwest: 3

DataFrame con 'sex' codificada:
        age  sex       bmi  children  smoker  region      charges
0  0.021739    0  0.321227         0       1       3  16884.92400
1  0.000000    1  0.479150         1       0       2   1725.55230
2  0.217391    1  0.458434         3       0       2   4449.46200
3  0.326087    1  0.181464         0       0       1  21984.47061
4  0.304348    1  0.347592         0       0       1   3866.85520


In [45]:
df.head()

#comprobamos los cambios que hemos hecho en el df para ver que los cambios que hemos hechos son correctos

age  sex       bmi  children  smoker  region      charges
0  0.021739    0  0.321227         0       1       3  16884.92400
1  0.000000    1  0.479150         1       0       2   1725.55230
2  0.217391    1  0.458434         3       0       2   4449.46200
3  0.326087    1  0.181464         0       0       1  21984.47061
4  0.304348    1  0.347592         0       0       1   3866.85520

In [57]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Supongamos que tienes un DataFrame df con tus datos

# Separar las características (X) y la variable objetivo (y)
X = df.drop(columns=['charges'])  # Características
y = df['charges']  # Variable objetivo

# Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Crear un modelo de regresión lineal
model = LinearRegression()

# Entrenar el modelo con los datos de entrenamiento
model.fit(X_train, y_train)

# Predecir los valores de la variable objetivo en el conjunto de prueba
y_pred = model.predict(X_test)

# Calcular métricas de evaluación
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Error cuadrático medio (MSE):", mse )
print("Coeficiente de determinación (R^2):", r2)





Error cuadrático medio (MSE): 35493102.6116505
Coeficiente de determinación (R^2): 0.8068466322629113


Vemos que el MSE es muy elevado, lo que nos dice que hay un error muy considerable a la hora de predecir.
El R2 nos dice que las variables x que tenemos explican alrededor del 80% de la variable y que queremos predecir

In [58]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Supongamos que tienes un DataFrame df con tus datos

# Separar las características (X) y la variable objetivo (y)
X = df.drop(columns=['charges'])  # Características
y = df['charges']  # Variable objetivo

# Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Crear un modelo de Gradient Boosting para regresión
model = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, random_state=42)

# Entrenar el modelo con los datos de entrenamiento
model.fit(X_train, y_train)

# Predecir los valores de la variable objetivo en el conjunto de prueba
y_pred = model.predict(X_test)

# Calcular métricas de evaluación
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Error cuadrático medio (MSE):", mse)
print("Coeficiente de determinación (R^2):", r2)


Error cuadrático medio (MSE): 18170259.107231136
Coeficiente de determinación (R^2): 0.901117499430293


Este modelo es capaz de mejorar ambas métricas respecto al anterior, el MSE es mas bajo, por lo que el error es menor y el R2 es del 90%.

In [59]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Supongamos que tienes un DataFrame df con tus datos

# Separar las características (X) y la variable objetivo (y)
X = df.drop(columns=['charges'])  # Características
y = df['charges']  # Variable objetivo

# Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Crear un modelo de árbol de decisión para regresión
model = DecisionTreeRegressor(random_state=42)

# Entrenar el modelo con los datos de entrenamiento
model.fit(X_train, y_train)

# Predecir los valores de la variable objetivo en el conjunto de prueba
y_pred = model.predict(X_test)

# Calcular métricas de evaluación
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Error cuadrático medio (MSE):", mse)
print("Coeficiente de determinación (R^2):", r2)


Error cuadrático medio (MSE): 40184326.610889345
Coeficiente de determinación (R^2): 0.7813170040369302


Este modelo se queda corto en ambas métricas.

In [60]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Supongamos que tienes un DataFrame df con tus datos

# Separar las características (X) y la variable objetivo (y)
X = df.drop(columns=['charges'])  # Características
y = df['charges']  # Variable objetivo

# Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Crear modelos
model_lr = LinearRegression()
model_gb = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, random_state=42)
model_dt = DecisionTreeRegressor(random_state=42)

# Entrenar los modelos
model_lr.fit(X_train, y_train)
model_gb.fit(X_train, y_train)
model_dt.fit(X_train, y_train)

# Predecir los valores de la variable objetivo en el conjunto de prueba
y_pred_lr = model_lr.predict(X_test)
y_pred_gb = model_gb.predict(X_test)
y_pred_dt = model_dt.predict(X_test)

# Ensemble: Promedio de las predicciones de los tres modelos
y_pred_ensemble = (y_pred_lr + y_pred_gb + y_pred_dt) / 3

# Calcular métricas de evaluación para el ensemble
mse_ensemble = mean_squared_error(y_test, y_pred_ensemble)
r2_ensemble = r2_score(y_test, y_pred_ensemble)

print("Error cuadrático medio (MSE) del ensemble:", mse_ensemble)
print("Coeficiente de determinación (R^2) del ensemble:", r2_ensemble)


Error cuadrático medio (MSE) del ensemble: 22342602.15515447
Coeficiente de determinación (R^2) del ensemble: 0.8784116199280516


Las métricas del ensemble están bien, pero no son capaces de mejorar las métricas del gradient boosting

In [62]:
import pandas as pd
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.tree import DecisionTreeRegressor

# Supongamos que tienes un DataFrame df con tus datos

# Separar las características (X) y la variable objetivo (y)
X = df.drop(columns=['charges'])  # Características
y = df['charges']  # Variable objetivo

# Crear modelos
model_lr = LinearRegression()
model_gb = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, random_state=42)
model_dt = DecisionTreeRegressor(random_state=42)

# Realizar validación cruzada para los modelos individuales
scores_lr_mse = cross_val_score(model_lr, X, y, cv=5, scoring='neg_mean_squared_error')
scores_lr_r2 = cross_val_score(model_lr, X, y, cv=5, scoring='r2')

scores_gb_mse = cross_val_score(model_gb, X, y, cv=5, scoring='neg_mean_squared_error')
scores_gb_r2 = cross_val_score(model_gb, X, y, cv=5, scoring='r2')

scores_dt_mse = cross_val_score(model_dt, X, y, cv=5, scoring='neg_mean_squared_error')
scores_dt_r2 = cross_val_score(model_dt, X, y, cv=5, scoring='r2')

# Calcular el promedio de los puntajes y convertirlos a valores positivos (MSE)
mse_lr = -scores_lr_mse.mean()
mse_gb = -scores_gb_mse.mean()
mse_dt = -scores_dt_mse.mean()

r2_lr = scores_lr_r2.mean()
r2_gb = scores_gb_r2.mean()
r2_dt = scores_dt_r2.mean()

print("Regresión Lineal:")
print("  MSE:", mse_lr)
print("  R^2:", r2_lr)

print("Gradient Boosting:")
print("  MSE:", mse_gb)
print("  R^2:", r2_gb)

print("Árbol de Decisión:")
print("  MSE:", mse_dt)
print("  R^2:", r2_dt)

# Crear ensemble
ensemble_models = [('lr', model_lr), ('gb', model_gb), ('dt', model_dt)]

# Realizar validación cruzada para el ensemble
scores_ensemble_mse = []
scores_ensemble_r2 = []
for model_name, model in ensemble_models:
    scores_mse = cross_val_score(model, X, y, cv=5, scoring='neg_mean_squared_error')
    scores_r2 = cross_val_score(model, X, y, cv=5, scoring='r2')

    mse = -scores_mse.mean()
    r2 = scores_r2.mean()

    scores_ensemble_mse.append((model_name, mse))
    scores_ensemble_r2.append((model_name, r2))

# Calcular el promedio de los puntajes y convertirlos a valores positivos (MSE)
mse_ensemble = sum([score for _, score in scores_ensemble_mse]) / len(scores_ensemble_mse)
r2_ensemble = sum([score for _, score in scores_ensemble_r2]) / len(scores_ensemble_r2)

print("Ensemble de Regresión Lineal, Gradient Boosting y Árbol de Decisión:")
print("  MSE:", mse_ensemble)
print("  R^2:", r2_ensemble)


Regresión Lineal:
  MSE: 36884478.7403644
  R^2: 0.7471285330088384
Gradient Boosting:
  MSE: 20880663.585826676
  R^2: 0.855946110524863
Árbol de Decisión:
  MSE: 41993131.82810299
  R^2: 0.7107057042792355
Ensemble de Regresión Lineal, Gradient Boosting y Árbol de Decisión:
  MSE: 33252758.051431358
  R^2: 0.7712601159376457


La validación cruzada es una herramienta esencial para evaluar, comparar y mejorar modelos de aprendizaje automático, lo que nos permite tomar decisiones informadas sobre qué modelo usar y cómo ajustarlo.







# Conclusiones

Primero de todo hemos verificado que no hay filas repetidas en el dataset y si hay, las eliminamos.

Tras ello hemos normalizado las variables numéricas con el método MinMaxScaler para poner en escala todos los números y para las variables categóricas he utilizado el método LabelEnconder para darle un valor numérico a cada una de ellas.

Con todo ello, he probado con distintos modelos predictivos que pudiesen ayudarme a conseguir las mejores métricas escogidas (MSE Y R2).

El modelo que mejor ha funcionado es el Gradient Boosting : Error cuadrático medio (MSE): 18170259.107231136
Coeficiente de determinación (R^2): 0.901117499430293

